### DISCUSSION TITLE
Error when running rosdep and colcon

### DISCUSSION THREAD

**Discussion by dSofikitis** [gstavrinos/ros2_kart_racing#21] [Q&A] [answered]
Hello there,

This is my ros2 workspace's structure (ros2 dir is located in ~):
    ros2:
    |   src:
    |   |   pub:
    |   |   |   include:
    |   |   |   |   pub: (empty)
    |   |   |   src:
    |   |   |   |   pub.cpp
    |   |   |   CMakeLists.txt
    |   |   |   package.xml
    |   |   sub:
    |   |   |   include:
    |   |   |   |   sub: (empty)
    |   |   |   src:
    |   |   |   |   sub.cpp
    |   |   |   CMakeLists.txt
    |   |   |   package.xml
    |   CMakeLists.txt
    |   package.xml

I'm having a hard time while trying to figure out the following:

When running `rosdep install -i --from-path src --rosdistro humble -y` in `~/ros2` directory I'm getting the following message: **#All Required rosdeps installed successfully**. Yet, nothing seems to actually happen. Then, when running `colcon build --symlink-install` in the same directory I'm getting the following error:

[0.193s] ERROR:colcon.colcon_core.package_identification:Exception in package identification extension 'cmake' in '.': Failed to extract project name from 'CMakeLists.txt'
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/colcon_core/package_identification/__init__.py", line 142, in _identify
    retval = extension.identify(_reused_descriptor_instance)
  File "/usr/lib/python3/dist-packages/colcon_cmake/package_identification/cmake.py", line 36, in identify
    raise RuntimeError(
RuntimeError: Failed to extract project name from 'CMakeLists.txt'

Starting >>> pub     
Starting >>> sub
--- stderr: pub                                                                
/home/ds/ros2/src/pub/src/pub.cpp:1:10: fatal error: rclcpp/rclcpp.hpp: No such file or directory
    1 | #include "rclcpp/rclcpp.hpp"
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
gmake[2]: *** [CMakeFiles/pub.dir/build.make:76: CMakeFiles/pub.dir/src/pub.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:137: CMakeFiles/pub.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
---
Failed   <<< pub [0.63s, exited with code 2]
Aborted  <<< sub [0.62s]

Summary: 0 packages finished [0.76s]
  1 package failed: pub
  1 package aborted: sub
  2 packages had stderr output: pub sub

What should I do?

**Comment by gstavrinos** [comment]
First things first: As we discussed, ROS packages should have unique names, thus `pub` and `sub` are not very good package names. You could name your nodes pub/sub though. A second thing that could be a problem is that you are creating two different packages for the publisher and subscriber. Check the exercise requirements again. (hint: You need a single node that **both** publishes and subscribes to different topics)

Continuing with your questions:

1. rosdep is running fine, there are no **specified** dependencies that it couldn't find (this doesn't mean that you are not missing non-specified dependencies).
2. colcon fails because there is something wrong with your packages. As stated in the error, there is something wrong with your CMakeLists. It specifically says that it cannot find the project name tag. Check a stable package for a non-malformed CMakeLists. Some tags are not optional (like the project name)

This problem is not due to a bad ROS installation though (👍), so keep focusing on how to implement your packages (and code) correctly. The provided examples should be more than enough.

**Comment by dSofikitis** [comment]
So, first I have to put both sub.cpp and pub.cpp in the same package.
(Note: pub.cpp simply sends an Empty message to the go_go_go topic)
And then, I should rename the package (just in case).

In the CMakeLists.txt and package.xml I copied and pasted the code that you provided in slide 31.

**Reply by gstavrinos** [reply]
You need a single node. Creating separate files for publishing and subscribing is your call, but since this node is really, really small, I would advise against it.

The slides do not provide complete solutions, ready for copy-pasting. Check out the correct way of creating packages. You can find it online and in the slides. You can also check out the provided ros2 examples (official ros2 examples mentioned in the slides as a url).

**Comment by dSofikitis** [comment, marked as answer]
Thanks a lot for your feedback!

Turns out the problem had nothing to do with anything that I mentioned above.

The problem has now been solved and everything runs as expected.
